### PR TITLE
new(userspace/libsinsp): add building blocks for filter field transformers

### PR DESCRIPTION
**Pull request by jasondellaluce**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area libsinsp

/area tests

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

As part of https://github.com/falcosecurity/libs/issues/1789, this PR adds the base constructs through which field transformers will be implemented in the first rollout of the proposed feature. This does not wire any of the new functionalities to the current filter field classes, but rather just adding an implementation to be used in subsequent changes.

This implements the `toupper`, `tolower`, and `b64` operations as defined in the proposal. There will be an opportunity to add more later on through the same building blocks.

Tests have been added for decent coverage about the bare functionality and the type consistency guarantees.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: userspace/libsinsp/sinsp_filter_transformer.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	uint8_t* ptr = nullptr;
+	uint32_t len = 0;
+};
+
+enum filter_transformer_type: uint8_t
+{
+	FTR_TOUPPER = 0,
+	FTR_TOLOWER = 1,
+	FTR_BASE64 = 2,
+	FTR_STORAGE = 3, // This transformer is only used internally
```
Comment (Andreagit97) [inline userspace/libsinsp/sinsp_filter_transformer.h:35]: for the reviewers `FTR_STORAGE` is just used internally for plugins, more details in the next PRs

**Review by Andreagit97 — DISMISSED** [review]
Amazing work, thank you for the cleanup!
/approve

**Comment by jasondellaluce** [comment]
/hold

I'm doing some research about the license of the base64 decoding code portion.

**Review by FedeDP — DISMISSED** [review]
Love this!
/approve

**Comment by FedeDP** [comment]
Also, please hold this until 0.16.x gets branched ;)

**Comment by Andreagit97** [comment]
if the license is ok i think we can also merge it since it is add only

**Comment by jasondellaluce** [comment]
I'm proposing using the following base64 library: https://github.com/tkislan/base64

Seems to be MIT licensed, has a decent amount of stars, and also suits our fine-grain allocation control needs for performance optimizations. Switched it from the previous one, and got tests to succeed locally.

**Comment by jasondellaluce** [comment]
> I'm proposing using the following base64 library: https://github.com/tkislan/base64
> 
> Seems to be MIT licensed, has a decent amount of stars, and also suits our fine-grain allocation control needs for performance optimizations. Switched it from the previous one, and got tests to succeed locally.

On a second thought, after discussing with @leogr, our best option is to use the same base64 library we used to employ in the past: https://raw.githubusercontent.com/istio/proxy/1.18.2/extensions/common/wasm/base64.h

Not only more maintained and employed in another CNCF project, but the licensing should allow us to add modifications (by referencing the initial authors) to fine-tune performance and allocations depending on our needs.

Will push the revised changes accordingly.

**Review by leogr — APPROVED** [review]
Great job :+1:

**Review by Andreagit97 — APPROVED** [review]
/approve

**Pull request merged**